# IPSA (CC01-1940): EDA y preprocesamiento para clasificación

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/juanestebancg2806/sugarcane-yield-quality-prediction/blob/main/01_eda_ipsa.ipynb)

Este notebook prepara los datos de la **parte de clasificación** del taller. Usa `BD_IPSA_1940.xlsx`, un
subconjunto del Ingenio Providencia que el propio Ingenio construyó con estas restricciones:

- **Variedad:** CC01-1940
- **Maduración:** química (madurante BONUS 250 EC)
- **Cosecha:** mecanizada en verde

Cada fila corresponde a una cosecha de una suerte (hacienda `FAZ` + suerte `TAL` en un `periodo`).
Las conclusiones aplican solo a suertes con estas características y **no describen al Ingenio completo**.
Este análisis es independiente del de regresión sobre el histórico de suertes
(`01_eda_suertes.ipynb` / `02_modelos_regresion.ipynb`).

**Objetivo:** predecir, antes de la cosecha, el nivel (bajo / medio / alto) de dos targets que se tratan
como problemas independientes:

- **TCH:** toneladas de caña por hectárea (cantidad)
- **Sacarosa:** % de sacarosa en caña (calidad)

**Flujo del notebook**

0. Alcance del dataset
1. Diccionario de datos
2. Chequeo de *leakage*
3. Auditoría de nulos disfrazados
4. Selección de predictores
5. Análisis de los dos targets
6. Relación predictor → target
7. VIF
8. Definición de las clases (terciles vs. umbrales de negocio)
9. Exportación y limitaciones

**Siguiente paso:** los modelos (logística multinomial regularizada y KNN) están en
`02_modelos_clasificacion.ipynb`.

> **Datos:** `BD_IPSA_1940.xlsx` no está en el repositorio porque el enunciado prohíbe publicar los datos.
> En Colab hay que subirlo manualmente al directorio de trabajo antes de ejecutar.

In [11]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm
from scipy import stats



pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)
pd.set_option("display.width", None)

sns.set_theme(style="whitegrid", context="notebook")
plt.rcParams["figure.figsize"] = (12, 6)
plt.rcParams["axes.titlesize"] = 14
plt.rcParams["axes.labelsize"] = 11

%matplotlib inline

## Carga de datos

Se lee la hoja `BD_IPSA` del archivo `BD_IPSA_1940.xlsx`, se elimina el índice residual que trae el
Excel y se revisa una vista general: primeras filas, tipos de dato, conteo de nulos y estadísticos
descriptivos.

In [12]:
# Lectura del dataset IPSA
df_ipsa = pd.read_excel("BD_IPSA_1940.xlsx", sheet_name="BD_IPSA")

# La primera columna de IPSA es un índice residual del Excel
if "Unnamed: 0" in df_ipsa.columns:
    df_ipsa = df_ipsa.drop(columns=["Unnamed: 0"])

print("BD_IPSA:", df_ipsa.shape)


BD_IPSA: (2187, 20)


In [13]:
# Vista rápida: IPSA (variedad CC01-1940)
display(df_ipsa.head())
df_ipsa.info()
display(df_ipsa.describe(include="all").T)

,NOME,FAZ,TAL,tipocorte,variedad,madurada,producto,dosismad,semsmad,edad,cortes,me,vejez,sacarosa,mes,periodo,TCH,lluvias,grupo_tenencia,pct_diatrea
0,AMAIME SILCA,81291,40,Mecanizado Verde,CC01-1940,SI,BONUS 250 EC REGULADOR FISIOLÓGICO,0.8,8.3,12.3,4,12.7,2.4,14.0,12,202012,112,137,3,6.2
1,AMAIME SILCA,81291,41,Mecanizado Verde,CC01-1940,SI,BONUS 250 EC REGULADOR FISIOLÓGICO,0.8,6.3,11.2,2,7.8,2.3,13.0,3,201903,157,0,3,3.5
2,AMAIME SILCA,81291,41,Mecanizado Verde,CC01-1940,SI,BONUS 250 EC REGULADOR FISIOLÓGICO,0.6,7.9,12.2,3,8.8,1.8,13.3,3,202003,167,68,3,4.3
3,AMAIME SILCA,81291,43,Mecanizado Verde,CC01-1940,SI,BONUS 250 EC REGULADOR FISIOLÓGICO,0.8,6.6,13.1,1,6.1,2.5,13.4,3,201903,156,0,3,3.5
4,AMAIME SILCA,81291,43,Mecanizado Verde,CC01-1940,SI,BONUS 250 EC REGULADOR FISIOLÓGICO,0.6,8.1,12.2,2,7.9,2.1,14.0,3,202003,151,68,3,4.3


<class 'pandas.DataFrame'>
RangeIndex: 2187 entries, 0 to 2186
Data columns (total 20 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   NOME            2187 non-null   str    
 1   FAZ             2187 non-null   int64  
 2   TAL             2187 non-null   object 
 3   tipocorte       2187 non-null   str    
 4   variedad        2187 non-null   str    
 5   madurada        2187 non-null   str    
 6   producto        2187 non-null   str    
 7   dosismad        2187 non-null   float64
 8   semsmad         2187 non-null   float64
 9   edad            2187 non-null   float64
 10  cortes          2187 non-null   int64  
 11  me              2187 non-null   float64
 12  vejez           2187 non-null   float64
 13  sacarosa        2187 non-null   float64
 14  mes             2187 non-null   int64  
 15  periodo         2187 non-null   int64  
 16  TCH             2187 non-null   int64  
 17  lluvias         2187 non-null   int64  
 18 

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
NOME,2187,285,SAN MIGUEL CARVAJAL,101,NaN,NaN,NaN,NaN,NaN,NaN,NaN
FAZ,2187.0,NaN,NaN,NaN,80588.332876,572.818299,80100.0,80222.0,80396.0,80660.0,82519.0
TAL,2187.0,273.0,1.0,258.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN
tipocorte,2187,1,Mecanizado Verde,2187,NaN,NaN,NaN,NaN,NaN,NaN,NaN
variedad,2187,1,CC01-1940,2187,NaN,NaN,NaN,NaN,NaN,NaN,NaN
madurada,2187,1,SI,2187,NaN,NaN,NaN,NaN,NaN,NaN,NaN
producto,2187,1,BONUS 250 EC REGULADOR FISIOLÓGICO,2187,NaN,NaN,NaN,NaN,NaN,NaN,NaN
dosismad,2187.0,NaN,NaN,NaN,0.993278,0.309096,0.0,0.8,1.0,1.2,9.0
semsmad,2187.0,NaN,NaN,NaN,9.164838,3.441579,-1.6,7.1,8.7,10.6,45.0
edad,2187.0,NaN,NaN,NaN,12.766118,1.117866,10.3,12.0,12.5,13.3,21.1


## 0. Alcance del dataset

`BD_IPSA_1940` es un recorte del Ingenio Providencia, no una muestra aleatoria. Si una condición de
manejo (variedad, maduración o tipo de cosecha) es igual en todas las filas, su efecto no se puede
estimar y las conclusiones solo aplican a suertes que compartan esa condición.

En esta sección se verifica si `variedad`, `madurada`, `producto` y `tipocorte` tienen un único valor
en todo el dataset.

In [14]:
# 0. Verificar que las columnas de alcance son constantes
cols_alcance = ["variedad", "madurada", "producto", "tipocorte"]

resumen_alcance = pd.DataFrame({
    "n_valores_unicos": df_ipsa[cols_alcance].nunique(),
    "valor": [df_ipsa[c].iloc[0] for c in cols_alcance],
})
display(resumen_alcance)

# Detener la ejecución si alguna columna de alcance no es constante
assert (resumen_alcance["n_valores_unicos"] == 1).all(), "Alguna columna de alcance NO es constante"

,n_valores_unicos,valor
variedad,1,CC01-1940
madurada,1,SI
producto,1,BONUS 250 EC REGULADOR FISIOLÓGICO
tipocorte,1,Mecanizado Verde


### Conclusión

Las cuatro columnas tienen un único valor en las 2.187 filas:

| Columna | Valor único | Qué no se puede estudiar con este dataset |
|---|---|---|
| `variedad` | CC01-1940 | Diferencias entre variedades |
| `madurada` | SI | Suertes maduradas vs. no maduradas |
| `producto` | BONUS 250 EC | Diferencias entre madurantes |
| `tipocorte` | Mecanizado Verde | Cosecha mecanizada vs. manual o con quema |

- Las cuatro columnas se excluyen como predictores porque, al ser constantes, no aportan información.
- Los resultados de los modelos aplican solo a suertes de CC01-1940 maduradas químicamente y
  cosechadas en verde con máquina.

## 1. Diccionario de datos

El diccionario oficial de `BD_IPSA` describe cada columna de forma breve y, en algunos casos, genérica.
Antes de usar una columna como predictor hay que confirmar qué mide y en qué unidad, porque una
interpretación equivocada puede invalidar el análisis posterior.

En esta sección se compara la descripción oficial de cada columna con su comportamiento real en los
datos (tipo, número de valores distintos y rango). Las columnas cuya descripción falte, sea ambigua o
no sea coherente con los valores observados se investigan por separado.

In [15]:
# Descripciones del diccionario oficial de BD_IPSA ("" = sin descripción)
dicc_oficial = {
    "NOME": "Nombre del registro (finca, campo o unidad de análisis)",
    "FAZ": "Código o identificador de la hacienda o finca",
    "TAL": "Identificador de una subunidad o lote",
    "tipocorte": "Tipo de corte realizado",
    "variedad": "Variedad del cultivo",
    "madurada": "Indicador relacionado con la maduración del cultivo",
    "producto": "",
    "dosismad": "Dosis aplicada de fertilizantes o agroquímicos",
    "semsmad": "Dosis aplicada de fertilizantes o agroquímicos",
    "edad": "Edad del cultivo al momento del registro (días, semanas o meses)",
    "cortes": "Número de cortes o cosechas realizadas en el periodo",
    "me": "",
    "vejez": "Edad o antigüedad del cultivo",
    "sacarosa": "Porcentaje o cantidad de sacarosa",
    "mes": "Mes del registro o de la cosecha",
    "periodo": "Periodo o año del registro",
    "TCH": "Toneladas de caña por hectárea",
    "lluvias": "Lluvias registradas en el periodo (mm)",
    "grupo_tenencia": "Categoría de tenencia del terreno",
    "pct_diatrea": "Porcentaje de infestación por diatrea",
}

filas = []
for col in df_ipsa.columns:
    s = df_ipsa[col]
    es_num = pd.api.types.is_numeric_dtype(s)
    filas.append({
        "columna": col,
        "descripcion_oficial": dicc_oficial.get(col, ""),
        "tipo": str(s.dtype),
        "n_unicos": s.nunique(),
        "min": s.min() if es_num else None,
        "mediana": s.median() if es_num else None,
        "max": s.max() if es_num else None,
        "ejemplos": list(s.drop_duplicates().head(4)),
    })

inventario = pd.DataFrame(filas).set_index("columna")
display(inventario)

,descripcion_oficial,tipo,n_unicos,min,mediana,max,ejemplos
columna,,,,,,,
NOME,"Nombre del registro (finca, campo o unidad de ...",str,285,NaN,NaN,NaN,"[AMAIME SILCA, ARANJUEZ, AURORA CUCALON, BARCE..."
FAZ,Código o identificador de la hacienda o finca,int64,285,80100.0,80396.0,82519.0,"[81291, 80552, 80601, 80492]"
TAL,Identificador de una subunidad o lote,object,273,NaN,NaN,NaN,"[40, 41, 43, 1]"
tipocorte,Tipo de corte realizado,str,1,NaN,NaN,NaN,[Mecanizado Verde]
variedad,Variedad del cultivo,str,1,NaN,NaN,NaN,[CC01-1940]
madurada,Indicador relacionado con la maduración del cu...,str,1,NaN,NaN,NaN,[SI]
producto,,str,1,NaN,NaN,NaN,[BONUS 250 EC REGULADOR FISIOLÓGICO]
dosismad,Dosis aplicada de fertilizantes o agroquímicos,float64,18,0.0,1.0,9.0,"[0.8, 0.6, 1.2, 1.0]"
semsmad,Dosis aplicada de fertilizantes o agroquímicos,float64,150,-1.6,8.7,45.0,"[8.3, 6.3, 7.9, 6.6]"


### Observaciones sobre el inventario

**Columnas cuya descripción es coherente con los valores observados:**

- `NOME` y `FAZ` tienen 285 valores distintos cada una, lo que sugiere que son el nombre y el código de
  la misma hacienda. `TAL` tiene 273 valores y es de tipo `object`, es decir, mezcla tipos de dato.
- `edad` va de 10.3 a 21.1 (mediana 12.5). La descripción oficial no fija la unidad, pero ese rango solo
  es coherente con meses: un ciclo de cultivo dura aproximadamente entre 12 y 18 meses.
- `cortes` toma valores enteros de 1 a 14.
- `sacarosa` (9.2 a 16.0 %), `TCH` (6 a 249 t/ha), `pct_diatrea` (0.2 a 25.5 %) y `lluvias` (0 a 1.468 mm)
  tienen unidades explícitas y rangos plausibles. Sus valores extremos se revisan más adelante.
- `mes` va de 1 a 12, y `periodo` tiene formato AAAAMM (201407 a 202101).

**Columnas que requieren revisión:**

| Columna | Problema |
|---|---|
| `me` | No tiene descripción en el diccionario. Rango de 3.4 a 15.0. |
| `vejez` | La descripción oficial ("antigüedad del cultivo") no es coherente con un rango de 0.2 a 102.9 con mediana 2.6, y la edad del cultivo ya está en `edad`. |
| `dosismad` | Descripción genérica y sin unidad. Máximo de 9.0 frente a una mediana de 1.0, y mínimo de 0.0 aunque todas las suertes figuran como maduradas. |
| `semsmad` | La descripción es idéntica a la de `dosismad`. Su rango (−1.6 a 45.0) no es coherente con una dosis, y un valor negativo no es coherente con ninguna cantidad física. |
| `grupo_tenencia` | Tres códigos (1, 2, 3) sin significado documentado. |

Además, antes de interpretar las columnas hay que confirmar qué representa cada fila. Con 285 haciendas
y 2.187 filas, cada hacienda aparece en varios registros.

### 1.1 Unidad de observación

Se verifica si `NOME` y `FAZ` identifican la misma hacienda, qué tipos de valores contiene `TAL` y si
sus valores de texto son códigos o números almacenados como texto, si `mes` es redundante con
`periodo`, y si la combinación hacienda + suerte + periodo identifica una fila única.

In [16]:
# 1.1 Unidad de observación

# ¿NOME y FAZ identifican la misma hacienda?
print("FAZ con más de un NOME:", (df_ipsa.groupby("FAZ")["NOME"].nunique() > 1).sum())
print("NOME con más de un FAZ:", (df_ipsa.groupby("NOME")["FAZ"].nunique() > 1).sum())

# TAL es de tipo object: ¿qué tipos de valores contiene?
display(df_ipsa["TAL"].map(lambda v: type(v).__name__).value_counts().to_frame("n_filas"))

# ¿mes coincide con los dos últimos dígitos de periodo?
print("Filas donde mes != periodo % 100:", (df_ipsa["mes"] != df_ipsa["periodo"] % 100).sum())

# TAL como texto para poder agrupar sin errores por la mezcla de tipos
tal_txt = df_ipsa["TAL"].astype(str)

# ¿hacienda + suerte + periodo identifica una fila única?
dup = df_ipsa.assign(TAL=tal_txt).duplicated(subset=["FAZ", "TAL", "periodo"], keep=False)
print("Filas con FAZ + TAL + periodo repetido:", dup.sum())

# ¿Cuántas cosechas tiene cada suerte en el dataset?
cosechas_por_suerte = df_ipsa.assign(TAL=tal_txt).groupby(["FAZ", "TAL"]).size()
print("Suertes distintas (FAZ + TAL):", len(cosechas_por_suerte))
display(cosechas_por_suerte.value_counts().sort_index().rename("n_suertes").to_frame())

FAZ con más de un NOME: 0
NOME con más de un FAZ: 0


,n_filas
TAL,
int,1598
str,589


Filas donde mes != periodo % 100: 0
Filas con FAZ + TAL + periodo repetido: 0
Suertes distintas (FAZ + TAL): 1115


,n_suertes
1,486
2,337
3,168
4,98
5,25
6,1


In [17]:
# Valores de TAL almacenados como texto: ¿son códigos alfanuméricos o números guardados como texto?
tal_str = df_ipsa.loc[df_ipsa["TAL"].map(lambda v: isinstance(v, str)), "TAL"]

es_numerico = tal_str.str.strip().str.fullmatch(r"\d+")
print("Valores de texto que son solo dígitos:", es_numerico.sum(), "de", len(tal_str))
print("Ejemplos:", tal_str.drop_duplicates().head(15).tolist())

Valores de texto que son solo dígitos: 0 de 589
Ejemplos: ['990A', '991B', '991C', '991D', '991E', '991G', '320B', '322A', '324A', '005A', '006A', '010A', '004A', '001A', '002B']


#### Conclusión de 1.1

- `NOME` y `FAZ` están en relación 1 a 1: son el nombre y el código de la misma hacienda, así que
  `NOME` es redundante con `FAZ`.
- `TAL` combina 1.598 valores enteros y 589 códigos alfanuméricos (por ejemplo, `005A` y `991B`).
  Ninguno de los valores de texto es un número almacenado como texto, por lo que la mezcla de tipos no
  duplica suertes. La letra final parece indicar una subdivisión de la suerte. Para agrupar, `TAL` se
  trata como texto.
- `mes` coincide en todas las filas con los dos últimos dígitos de `periodo`, así que su información ya
  está contenida en `periodo`.
- La combinación `FAZ` + `TAL` + `periodo` no se repite, así que **cada fila es una cosecha de una
  suerte**.
- El dataset contiene 1.115 suertes distintas. 486 aparecen en una sola cosecha y 629 en dos o más
  (hasta seis):

| Cosechas por suerte | 1 | 2 | 3 | 4 | 5 | 6 |
|---|---|---|---|---|---|---|
| Número de suertes | 486 | 337 | 168 | 98 | 25 | 1 |

Como una misma suerte aparece en varias filas, las observaciones no son independientes entre sí. Una
partición aleatoria de filas en entrenamiento y prueba puede dejar cosechas de la misma suerte en
ambos conjuntos, y el desempeño en prueba resultaría optimista. Esto se tiene en cuenta al construir
la partición en `02_modelos_clasificacion.ipynb`.

### 1.2 Columna `vejez`

El diccionario oficial describe `vejez` como la edad o antigüedad del cultivo. La guía del taller, con
información validada por el Ingenio, la define como las horas transcurridas entre el corte (o una quema
accidental) y la llegada de la caña al molino. Según esa fuente, en operación normal el valor es menor
a 10 horas, y valores del orden de 70 horas corresponden a quemas accidentales.

Se revisa cuál de las dos interpretaciones es coherente con la distribución observada, comparando
`vejez` con `edad`, que ya representa la edad del cultivo en meses.

In [18]:
# 1.2 Columna vejez
vejez = df_ipsa["vejez"]

display(vejez.describe(percentiles=[.25, .5, .75, .90, .95, .99]).round(2).to_frame("vejez"))

# Si vejez fuera la edad del cultivo, debería parecerse a edad (10-21 meses) y correlacionar con ella
print("Correlación de Spearman vejez-edad:", round(vejez.corr(df_ipsa["edad"], method="spearman"), 3))

# Proporción de registros por rangos de horas (referencia de la guía: < 10 h en operación normal)
rangos = pd.cut(vejez, bins=[0, 10, 24, 48, np.inf], labels=["<10 h", "10-24 h", "24-48 h", ">48 h"])
display(rangos.value_counts().sort_index().to_frame("n_filas")
        .assign(pct=lambda d: (d["n_filas"] / len(df_ipsa) * 100).round(1)))

,vejez
count,2187.00
mean,4.17
std,6.34
min,0.20
25%,2.10
50%,2.60
75%,3.30
90%,6.30
95%,14.47
99%,33.57


Correlación de Spearman vejez-edad: 0.112


,n_filas,pct
vejez,,
<10 h,2038,93.2
10-24 h,108,4.9
24-48 h,32,1.5
>48 h,9,0.4


#### Conclusión de 1.2

- La correlación de Spearman entre `vejez` y `edad` es de 0.112. Si `vejez` representara la antigüedad
  del cultivo, estaría fuertemente relacionada con `edad`, así que la descripción del diccionario
  oficial no corresponde a esta columna.
- La distribución es coherente con horas entre corte y molienda: mediana de 2.6 h, el 75 % de los
  registros por debajo de 3.3 h y el 93.2 % por debajo de 10 h, el umbral de operación normal que
  indica la guía.
- Hay una cola larga pequeña: 108 registros (4.9 %) entre 10 y 24 h, 32 (1.5 %) entre 24 y 48 h y
  9 (0.4 %) por encima de 48 h, con un máximo de 102.9 h. Según la guía, estos valores corresponden a
  eventos operativos como quemas accidentales, no a errores de escala. Su tratamiento se decide al
  revisar valores extremos.
- Se adopta la definición de la guía: **`vejez` = horas entre el corte de la caña y su llegada al
  molino**. Al ser una medida posterior al corte, su uso como predictor se evalúa en la sección 2.

### 1.3 Columna `me`

Ninguna fuente describe esta columna. Su rango observado va de 3.4 a 15.0, con mediana 9.1.

**Hipótesis principal: materia extraña (% en peso).** Es el material que llega al molino junto con los
tallos y no es caña aprovechable: hojas, cogollos, chulquines y tierra. Hay tres argumentos a favor:

- El rango es coherente con los porcentajes que reporta Cenicaña, por ejemplo un promedio de 7.48 % en
  Riopaila [1]. En cosecha mecanizada en verde se esperan valores mayores, porque la máquina arrastra
  más hojas y cogollos.
- La materia extraña y el tiempo entre corte y molienda (`vejez`) se miden juntos al recibir la caña
  en el molino [2].
- La materia extraña no contiene sacarosa y la diluye. Cenicaña reporta una reducción de 0.13 a 0.17
  puntos de pol por cada 1 % de materia extraña [1]. Por eso se espera una relación negativa con
  `sacarosa`.

**Hipótesis alternativa: un tiempo en meses**, por ejemplo la edad del cultivo al aplicar el madurante.
Se descarta si `me` supera a `edad` en una proporción relevante de filas.

**Fuentes**

1. Cenicaña. *Calidad de la caña de azúcar*. En: El cultivo de la caña en la zona azucarera de Colombia,
   pp. 337-354. https://www.cenicana.org/pdf_privado/documentos_no_seriados/libro_el_cultivo_cana/libro_p337-354.pdf
2. Universidad Nacional Abierta y a Distancia (UNAD). *Impacto generado por la presencia de materia
   extraña y tiempo de permanencia de caña de azúcar cosechada en campo en la disminución del contenido
   de sacarosa en los ingenios del Valle del Cauca*. https://repository.unad.edu.co/handle/10596/51320

In [19]:
# 1.3 Columna me
me = df_ipsa["me"]

# Hipótesis alternativa: si me > edad en muchas filas, me no puede ser un tiempo en meses previo al corte
n_mayor = (me > df_ipsa["edad"]).sum()
print(f"Filas con me > edad: {n_mayor} de {len(df_ipsa)} ({n_mayor / len(df_ipsa):.1%})")

display(me.describe(percentiles=[.01, .05, .25, .5, .75, .95, .99]).round(2).to_frame("me"))

# Spearman: captura relaciones monótonas aunque no sean lineales
cols = ["sacarosa", "TCH", "vejez", "edad", "cortes", "lluvias", "mes"]
rho = df_ipsa[cols + ["me"]].corr(method="spearman")["me"].drop("me")
display(rho.round(3).to_frame("rho_spearman_con_me"))

# Evolución por año de cosecha
display(df_ipsa.groupby(df_ipsa["periodo"] // 100)["me"]
        .agg(["count", "median", "mean", "min", "max"]).round(2))

Filas con me > edad: 83 de 2187 (3.8%)


,me
count,2187.00
mean,9.20
std,1.57
min,3.40
1%,5.70
5%,6.80
25%,8.20
50%,9.10
75%,10.10
95%,12.00


,rho_spearman_con_me
sacarosa,-0.285
TCH,0.017
vejez,-0.044
edad,-0.129
cortes,0.166
lluvias,0.004
mes,-0.061


,count,median,mean,min,max
periodo,,,,,
2014,1,5.8,5.80,5.8,5.8
2015,35,7.7,7.88,5.4,10.3
2016,110,8.5,8.61,4.9,13.3
2017,217,9.1,9.06,4.4,13.2
2018,408,8.8,8.99,3.9,14.5
2019,602,9.1,9.31,5.0,15.0
2020,728,9.2,9.36,3.4,15.0
2021,86,9.6,9.68,6.6,12.7


#### Conclusión de 1.3

- **Hipótesis temporal descartada.** La correlación de Spearman entre `me` y `edad` es débil y negativa
  (−0.129). Si `me` fuera un tiempo en meses dentro del ciclo del cultivo, crecería con la edad. Además,
  en 83 filas (3.8 %) `me` supera a `edad`, algo imposible para un evento anterior al corte.
- **Hipótesis de materia extraña respaldada.** La relación más fuerte de `me` es con `sacarosa`
  (ρ = −0.285), negativa como se espera si la materia extraña diluye el contenido de sacarosa. Con `TCH`
  la relación es nula (ρ = 0.017).
- La distribución es compacta y coherente con un porcentaje en peso: el 90 % central de los registros
  está entre 6.8 y 12.0, con mediana 9.1.
- La mediana anual pasa de 7.7 en 2015 a 9.6 en 2021 (2014 tiene un solo registro). Esa tendencia es
  compatible con un indicador de la operación de cosecha más que con una propiedad del cultivo.
- Dos resultados no aportan evidencia a favor: la correlación con `lluvias` es prácticamente nula
  (ρ = 0.004), aunque `lluvias` tiene ceros cuya validez se revisa en la sección 3, y la correlación
  positiva con `cortes` (ρ = 0.166) no tiene una explicación establecida.
- Se adopta como hipótesis de trabajo, con confianza moderada y sin confirmación del Ingenio:
  **`me` = materia extraña (% en peso de la caña entregada)**. Bajo esta interpretación, `me` se mide al
  recibir la caña en el molino, así que su uso como predictor se evalúa en la sección 2.

### 1.4 Columna `dosismad`

El diccionario oficial describe `dosismad` de forma genérica ("dosis aplicada de fertilizantes o
agroquímicos") y no indica la unidad. Como `producto` es constante (BONUS 250 EC), la columna se
interpreta como la dosis del madurante.

BONUS 250 EC es un regulador de crecimiento cuyo ingrediente activo es el trinexapac-etil. Para uso
como madurante en caña, el fabricante recomienda 1.2 L/ha, equivalente a 7-8 cc por tonelada de caña
estimada al momento de la aplicación [1]. La mediana observada (1.0) es coherente con una dosis en L/ha.

Quedan dos aspectos por revisar:

- El valor máximo de 9.0 está muy por encima de la dosis recomendada en L/ha, pero dentro del orden
  de magnitud de la dosis expresada en cc por tonelada.
- Hay registros con dosis 0.0, aunque todas las suertes figuran como maduradas (`madurada = SI`).

**Fuentes**

1. Syngenta Colombia. *BONUS® 250 EC – Regulador de crecimiento*.
   https://www.syngenta.com.co/product/crop-protection/regulador-de-crecimiento/bonusr-250-ec

In [20]:
# 1.4 Columna dosismad
dosis = df_ipsa["dosismad"]

# Distribución completa de valores (la columna tiene pocos valores distintos)
display(dosis.value_counts().sort_index().to_frame("n_filas")
        .assign(pct=lambda d: (d["n_filas"] / len(df_ipsa) * 100).round(2)))

# Registros fuera del rango habitual: dosis 0 o por encima de 2.0 L/ha
cols_ver = ["FAZ", "TAL", "periodo", "dosismad", "semsmad", "edad", "sacarosa", "TCH"]
display(df_ipsa.loc[(dosis == 0) | (dosis > 2.0), cols_ver].sort_values("dosismad"))

,n_filas,pct
dosismad,,
0.0,1,0.05
0.2,2,0.09
0.3,4,0.18
0.5,103,4.71
0.6,56,2.56
0.7,12,0.55
0.8,757,34.61
1.0,492,22.50
1.1,57,2.61


,FAZ,TAL,periodo,dosismad,semsmad,edad,sacarosa,TCH
1204,80232,1,201603,0.0,8.1,12.9,13.3,157
1652,80179,1,201512,9.0,6.1,13.3,14.6,104
